configuration/pih/petl/jobs/hiv/viral_load/source.sql

```diff
@@ -0,0 +1,126 @@
+SET sql_safe_updates = 0;
+
+DROP TABLE IF EXISTS temp_hiv_constructs;
+DROP TABLE IF EXISTS temp_viral_load;
+DROP TABLE IF EXISTS temp_vl_sample_taken_date;
+DROP TABLE IF EXISTS temp_days_since_vl;
+
+-- hiv constructs table
+CREATE TEMPORARY TABLE temp_hiv_constructs
+(
+patient_id 				INT,
+encounter_id 			INT
+);
+
+INSERT INTO temp_hiv_constructs (patient_id, encounter_id)
+SELECT person_id, encounter_id FROM obs WHERE voided = 0 AND concept_id = CONCEPT_FROM_MAPPING("PIH", "HIV viral load construct") AND encounter_id IN (SELECT encounter_id FROM encounter
+WHERE encounter_type =(SELECT encounter_type_id FROM encounter_type WHERE name = "Laboratory Results"));
+
+-- specimen cpllection date table
+CREATE TEMPORARY TABLE temp_vl_sample_taken_date
+(
+encounter_id 			INT,
+vl_sample_taken_date	DATE
+);
+
+INSERT INTO temp_vl_sample_taken_date(encounter_id, vl_sample_taken_date)
+SELECT encounter_id, DATE(encounter_datetime) FROM encounter WHERE encounter_id IN (SELECT encounter_id FROM temp_hiv_constructs);
+
+-- viral load details table
+CREATE TABLE temp_viral_load
+(
+patient_id						INT,
+encounter_id					INT,
+vl_sample_taken_date_estimated 	VARCHAR(11),
+vl_result_date					DATE,
+vl_test_outcome					VARCHAR(255),
+vl_result_detectable			VARCHAR(255),
+viral_load						INT,
+vl_type							VARCHAR(50)
+);
+
+INSERT INTO temp_viral_load (patient_id, encounter_id)
+(SELECT patient_id, encounter_id FROM temp_hiv_constructs);
+
+UPDATE temp_viral_load SET vl_sample_taken_date_estimated =  OBS_VALUE_CODED_LIST(encounter_id, 'PIH', '11781', 'en'); 
+
+UPDATE temp_viral_load SET vl_result_date =  OBS_VALUE_DATETIME(encounter_id, 'PIH', 'DATE OF LABORATORY TEST');
+
+UPDATE temp_viral_load SET vl_result_detectable =  OBS_VALUE_CODED_LIST(encounter_id, 'CIEL', '1305', 'en');
+
+UPDATE temp_viral_load SET viral_load =  OBS_VALUE_NUMERIC(encounter_id, 'CIEL', '856');
+
+-- days since last viral load table
+CREATE TEMPORARY TABLE temp_days_since_vl
+(
+patient_id		INT,
+result_date		DATE
+);
+
+INSERT INTO temp_days_since_vl(patient_id, result_date)
+SELECT patient_id, MAX(vl_result_date) FROM temp_viral_load GROUP BY patient_id;
```

Comment:
Not sure of the business rules here, but shouldn't this be based on on the specimen collection date, not the result date?

Also, on a larger question, would this be the kind of thing better calculated in a downstream report by MEQ (ie in PowerBI, etc) rather than defined here?  This would make it easier for the MEQ to modify how it's calculated (and to be reminded how it's calculated, etc)

Interested in @dmdesimone 's thoughts as well.

Reply:
Ah, re: result date vs sample date, I see result date is the requirement from the Excel document.  Seems strange to  me, but maybe that's correct!

Reply:
defer to @dmdesimone about the calculations. I think this is what the excel file has

Reply:
Per a separate discussion with Debbie, we agreed that yes, this should be Sample (encounter) date!